src/Component.php

```diff
@@ -0,0 +1,223 @@
+   *
+   * @param $handle
+   *   The component handle to extract the parts for.
+   *   Example values:
+   *   - @foo/baz.twig
+   *   - @foo/baz--bar.twig
+   */
+  public function __construct(string $handle) {
+    list($this->pathname, $this->name, $this->variants) = $this->extractParts($handle);
+  }
+
+  /**
+   * Returns the default variables from the component's definition file.
+   *
+   * @return array
+   *   The default variables of the component.
+   */
+  public function getDefaultVariables(): array {
+    $component_definition = $this->loadDefinition($this->getPathname());
+
+    $defaults = [];
+    if (isset($component_definition['context'])) {
+      $defaults = $this->mergeContext($defaults, $component_definition['context']);
+    }
+    if (!empty($this->getVariants()) && isset($component_definition['variants'])) {
+      foreach ($this->getVariants() as $variant_modifier) {
+        if ($variant_defaults = $this->getVariantDefaultVariables($variant_modifier, $component_definition['variants'])) {
+          $defaults = $this->mergeContext($defaults, $variant_defaults);
+        }
+      }
+    }
+    return $defaults;
+  }
+
+  /**
+   * Merges a context default variables, with support for attributes.
+   *
+   * Note: The initial set of default variables should not be set directly to
+   * $base but should also merged with this helper function.
+   *
+   * @param array $base
+   *   The context default variables (result) array to merge into.
+   * @param array $override
+   *   The context variables of a certain scope; i.e., the components base
+   *   context variables or the context variables of a variant.
+   *
+   * @return array
+   *   The $base array merged with $override variables,
+   */
+  protected function mergeContext(array $base, array $override): array {
+    foreach ($override as $name => $value) {
+      // Any context variable containing "attributes" is considered a data array
+      // for Drupal Attributes; e.g., title_attributes, author_attributes, etc.
+      if (strpos($name, 'attributes') === FALSE) {
+        $base[$name] = $value;
+      }
+      else {
+        // Handle the special merging of classes first.
+        // Also ensure that the 'class' attribute on $base is always an array.
+        if (isset($value['class'])) {
+          $base[$name]['class'] = array_merge($base[$name]['class'] ?? [], is_array($value['class']) ? $value['class'] : explode(' ', $value['class']));
+          unset($value['class']);
+        }
+        if (!isset($base[$name])) {
+          $base[$name] = [];
+        }
+        $base[$name] = array_merge($base[$name], $value);
+      }
+    }
+    return $base;
+  }
+
+  /**
+   * Loads the Fractal YAML component definition file.
+   *
+   * @param string $pathname
+   *   The pathname of the component's template file.
+   *
+   * @return array
+   *   The parsed component definition.
+   */
+  protected function loadDefinition(string $pathname) {
```

Comment:
The return type declaration `: ?array` for mixed values (`NULL/VOID/array`) is initially available since PHP7.1. Should we add a return type declaration in even if it will cover the recent versions only?

Reply:
We should only require PHP 7.0 at this point.  (The requirement should be specified in the module's info.yml file.)